Modernize package w/ switch from {httr} to {httr2}

elipousson opened the pull request:
Here is the summary of major changes in the pull request from #49 

- Switched over to using {cli} for almost all messages and alerts (I left addDomainInfo.R and token.R alone)
- Added a `geometry` parameter to `esri2sf()` to begin the process of allowing other types of spatial filters (not just bounding boxes)
- Added the new `esriIndex()` function described in #39 
- Added a new `esriInfo()` function that includes a sitemap option that provides similar (but more limited) functionality as esriIndex
- Reorganized the code since zzz.R was getting very crowded (I split off esri2sfGeom.R and esriLayers.R)

elipousson committed:
refactor: split esri2sfGeom functions into separate file

elipousson committed:
feat: Add esriRequest function

elipousson committed:
feat: Add esriInfo function

elipousson committed:
feat: add esriCatalog, esriIndex, and esriIndexLayers functions

elipousson committed:
style: adjust whitespace w/ styler

elipousson committed:
refactor: switch esriLayers to use esriRequest

- refactor: switch esriLayers to use esriRequest (replacing httr w/ httr2)
- refactor: add is_urlServer helper function

elipousson committed:
refactor: update esri2sf and esri2df to work w/  httr2

- refactor: update esri2sf and esri2df to work w/  httr2
- feat: add geometry parameter to allow spatial filters other than bounding box objects  (supported by new sf2geometry helper)
- feat: add support to bbox parameter to convert sf objects to bbox (supported by new bbox2geometry helper)
- refactor: rename bbox parameter to geometry for getEsriFeatures and other helper functions
- refactor: switch replaceDomainInfo default to FALSE
- refactor: disable esriUrl_isValid check for esrimeta
- refactor: switch to use cli for messages and alerts
- docs: separate esrimeta docs from esri2sf

elipousson committed:
refactor: update functions to work w/ httr2 and cli

elipousson committed:
style: minor style changes

elipousson committed:
refactor: update NAMESPACE + DESCRIPTION

- refactor: update NAMESPACE to export new functions + update imports
- refactor: update DESCRIPTION to add magick and purrr to Suggests and cli, xml2 and httr2 to Imports

elipousson committed:
docs: update function documentation

elipousson committed:
fix: catch blank layer geometryType along w/ NULL geometryType

elipousson committed:
docs: swap example URLs in get started vignette

elipousson committed:
refactor: add user agent to esriRequest

elipousson committed:
refactor: add explicit imports for utilized functions

elipousson committed:
docs: add myself to contributor list and alphabetize Imports

elipousson committed:
refactor: add getLayerCRS helper function

- refactor: add getLayerCRS helper function
- refactor: switch default for outFields and where to NULL

elipousson committed:
fix: correct missing commas in DESCRIPTION

elipousson committed:
docs: add missing parameter definitions for esriRequest

elipousson committed:
refactor: add support to bbox objects to sf2geometry

- refactor: add support to bbox objects to sf2geometry (drops need for bbox2geometry helper)
- refactor: add sf2geometryType helper function
- refactor: add layerCRS as parameter to getLayerCRS function (only used to set default currently)
- refactor: set outFields to NULL for esri2df and token to NULL for esrimeta

elipousson committed:
fix: correct specification of `f` API parameter

- fix: correct specification of `f` API parameter (required for implementation of ImageServer download functions that require both an f and format parameter)
- docs: minor updates to multiple parameter definitions

elipousson commented:
I switched this pull request into a draft because I found a few critical issues (fixed now I think) and figured a more cautious approach would be better to make sure all outstanding issues are resolved before merging would be appropriate. Looking forward to your feedback @jacpete (or @yonghah if you have time to also take a look).

elipousson committed:
fix: add check for NULL values in field domain

- fix: add check for NULL values in field domain
- refactor: set token default to NULL

elipousson committed:
refactor: set token default to NULL

elipousson committed:
refactor: simplify handling of geometryType parameter if bbox is provided

elipousson committed:
fix: swap format parameter for f parameter for esriLayers

elipousson committed:
fix: correct parameter names in esriRequest

- fix: correct parameter names in esriRequest
- fix: use user-provided geomType if it is not NULL and warn if geomType does not match layerInfo$geometryType
- fix: update esri2sf to correctly handle missing spatialReference values if crs is not NULL
- fix: update extent2bbox to correctly reference min/max values

elipousson commented:
That latest commit improves support for layers with a missing spatial reference. Here is an example if you wanted to test/compare with the current version which cannot access this layer: https://services1.arcgis.com/mVFRs7NF4iFitgbY/arcgis/rest/services/21st_Century_Schools/FeatureServer/0

elipousson committed:
fix: correct misnamed variable in sf2geometry

- fix: correct misnamed variable in sf2geometry
- feat: convert sf objects to bbox objects if passed to bbox (but not to geometry)

elipousson committed:
refactor: improve handling of transform for bbox parameter

elipousson commented:
Argh. I introduced a bug with the geometry parameter at some point but I'm having difficulty tracking it down. Despite my earlier enthusiasm this isn't quite ready for review yet.

elipousson committed:
feat: add cache parameter to esriRequest

- feat: add cache parameter to esriRequest
- refactor: remove unused error check

elipousson committed:
refactor: move simplifyDataFrame parameter to getEsriFeaturesByIds

- refactor: move simplifyDataFrame parameter to getEsriFeaturesByIds and remove unused (I think) error check

elipousson committed:
docs: fill in missing info for esriLayers parameter definitions

elipousson committed:
fix: correct issues w/ bbox/geometry parameters

- fix: correct issues w/ bbox/geometry parameters
- refactor: add `simplifyVector = TRUE` parameter to `esrimeta()` for consistency w/ existing function

elipousson committed:
docs: update cache parameter to docs

jacpete commented:
@elipousson Sorry for the crazy wait on this. I am in the middle of transitioning jobs and I just haven't had any free time outside of work to take a look at this. I should be able to look at this in the evenings and get back to you next week. Just wanted to make sure you knew that I have seen this and it is on my schedule.

elipousson commented:
This is so frustrating. I swear I had this working but either httr2 changed or more likely I introduced a weird regression at some point and almost all of these commits on this branch don't seem to work at all. Let me know if you're encountering a similar issue or if it is unique to my set-up. If you're running into a similar issue, I honestly think I may need to start back at the beginning and bring over the refactored code bit by bit to make sure I can avoid a repeat.

elipousson commented:
Aha! I think the issue is at least in part related to the escaping of the commas separating the object id values in the query referenced in this issue https://github.com/r-lib/httr2/issues/152

There is an active pull request to fix this issue: https://github.com/r-lib/httr2/pull/153

I'll incorporate the new function once the pull request is accepted which should clarify what issues are with the refactored {esri2sf} and what issues are with {httr2}.

elipousson committed:
refactor: drop explicit returns at end of functions

elipousson committed:
refactor: prefix . to perform and cache parameters to avoid conflict w/ possible query parameters

elipousson committed:
docs: correct spelling error in parameter definitions

elipousson committed:
fix: do not encode objectIds or outFields

elipousson committed:
test: update snapshots

elipousson committed:
docs: update docs

elipousson committed:
refactor: add rappdirs to Suggests

elipousson committed:
test: update esriLayers snapshot

elipousson committed:
feat: add .max_seconds parameter to `esriRequest` for httr2::req_retry

elipousson committed:
style: run styler

elipousson committed:
refactor: add maxRecords parameter to getMaxRecordsCount

- refactor: add maxRecords parameter to getMaxRecordsCount
- refactor: switch returns to follow tidyverse style

elipousson committed:
refactor: add maxRecords and spatialRel parameters

- refactor: add maxRecords and spatialRel parameters
- refactor: simplify returns for consistency w/ tidyverse style

elipousson committed:
refactor: replace pbapply w/ cli_progress_along

elipousson committed:
refactor: replace askForPassword w/ readline

elipousson committed:
docs: update function docs

elipousson committed:
test: update esriLayers snapshot

elipousson committed:
docs: update README

elipousson committed:
refactor: tidy DESCRIPTION

elipousson commented:
OK – this may not be perfect but I added a maxRecords parameter that can be passed through all the way to the `getMaxRecordsCount()` helper function. All of the servers that were kicking back errors with the new version work fine if you provide a lower maxRecords value.

I suspect that the 1000 max records from the layer info may be a default value but the servers actually *aren't* set up to handle that many requests at once. If there is a more elegant solution, I'd be happy to hear it. At a minimum, I can add an error message that suggests using the maxRecords parameter if a request fails. Otherwise, I think the significant number of new features in this branch may make it worth integrating even if this isn't the most seamless solution.

elipousson committed:
feat: add error message that suggests using maxRecords

elipousson committed:
fix: correct typo in comment

elipousson committed:
refactor: improve error handling when maxRecords may be needed

elipousson committed:
docs: add codemeta.json

elipousson committed:
refactor: add keyring (used in tests) to Suggests

elipousson committed:
fix: correct issue w/ NULL geometryType

elipousson committed:
test: add tests for esri2sf + esriIndex

elipousson committed:
refactor: remove unused bbox2geometry function

elipousson committed:
refactor: update NAMESPACE

elipousson commented:
Ah. On reflection, the addition of the maxRecords parameters is an insufficient solution because it won't be backwards compatible with existing code (including a bunch of my wrapper functions for Baltimore City FeatureLayers that use esri2sf). Review for this pull request is still welcome but it isn't ready to merge yet. I'll add more thoughts to the issue.

elipousson committed:
refactor: don't import xml2 + remove rdname tag

elipousson committed:
refactor: replace getWKTidAuthority w/ st_crs()$srid

elipousson committed:
refactor: add bbox2geometry helper function

- refactor: add bbox2geometry and is_missing_geomType helper function, shorten line lengths, and use inherits for class checks
- refactor: replace getWKTidAuthority w/ st_crs()$srid
- docs: correct outdated parameter definition

elipousson committed:
refactor:adjust spacing + warnings + other minor changes

elipousson committed:
refactor: remove getWKTidAuthority

elipousson committed:
refactor: drop DBI + RSQLite Imports

elipousson committed:
fix: make select for relocate more explicit

- fix: make select for relocate more explicit in esriIndex
- style: shorten line lengths in esriIndex.R
- docs: minor updates to documentation

elipousson committed:
refactor: update esriRequest for URLs over 2048 chars

elipousson committed:
refactor: simplify returns + switch to use cli_inform

elipousson committed:
refactor: simplify returns + progress bar logic

elipousson committed:
fix: correct issue w/ missing outSR from query

elipousson committed:
feat: remove maxRecords parameter + modify crs default

elipousson committed:
refactor: update NAMESPACE + minor other changes

elipousson committed:
docs: update function docs

elipousson committed:
fix: getEsriFeatures must pass objectIds explicitly

- fix: getEsriFeatures and getObjectIds must pass objectIds explicitly
- fix: make sure layerInfo$type is not NULL before checking value in esri2df
- fix: esriIndex reverse breaking change on relocate function

elipousson committed:
test: remove esri2df second test (bad url) + update snapshot

elipousson committed:
docs: update README

- docs: update README (note there is a possible outstanding issue with the polyline example)

elipousson committed:
fix: make esriIndex work w/ even only folders

elipousson committed:
feat: add esrigeocode function

elipousson committed:
feat: add internal getbyIds param to getEsriFeatures

- feat: add internal getbyIds param to getEsriFeatures to support accessing features w/o the interim step of getObjectIds
- refactor: switch crs default to NULL for internal functions
- fix: correct issue w/ getEsriFeaturesByIds using id instead of objectIds parameter

The addition of the getbyIds parameter could be temporary or could be set as the new default behavior for queries below some size threshold.

elipousson committed:
fix: switch to use httr2::req_body_form

- fix: switch to use httr2::req_body_form instead of httr2::req_body_json which simplifies the code and avoids the need for the getbyIds parameter for getEsriFeatures (maxed out at 1000 features anyway)

elipousson committed:
docs: update README + codemeta.json

elipousson committed:
docs: expand README w/ additional functions

- docs: expand README w/ additional functions
- feat: add lowercase name version of esriindex and esricatalog

elipousson committed:
docs: correct example w/ bbox filter for table

elipousson committed:
feat: add first try for esri2rast function

elipousson committed:
refactor: validate format arg

- refactor: validate format arg and improve error message
- docs: expand documentation

elipousson committed:
feat: add quiet arg

- feat: add quiet arg to optionally suppress messages
- refactor: minor changes to messages

elipousson committed:
fix: correct issue w/ Group Layer services

elipousson commented:
I figure it makes sense to close this pull request for now – but happy to re-open it in the future.
